pkg/operator/options/options.go

```diff
@@ -0,0 +1,177 @@
+	fs.StringVar(&o.NetworkPolicy, "network-policy", env.WithDefaultString("NETWORK_POLICY", ""), "AKS cluster network policy.")
+	fs.Var(newNodeIdentitiesValue(env.WithDefaultString("NODE_IDENTITIES", ""), &o.NodeIdentities), "node-identities", "User assigned identities for nodes.")
+}
+
+func (o Options) GetAPIServerName() string {
+	endpoint, _ := url.Parse(o.ClusterEndpoint) // assume to already validated
+	return endpoint.Hostname()
+}
+
+func (o *Options) Parse(fs *coreoptions.FlagSet, args ...string) error {
+	if err := fs.Parse(args); err != nil {
+		if errors.Is(err, flag.ErrHelp) {
+			os.Exit(0)
+		}
+		return fmt.Errorf("parsing flags, %w", err)
+	}
+
+	// Check if each option has been set. This is a little brute force and better options might exist,
+	// but this only needs to be here for one version
+	o.setFlags = map[string]bool{}
+	cliFlags := sets.New[string]()
+	fs.Visit(func(f *flag.Flag) {
+		cliFlags.Insert(f.Name)
+	})
+	fs.VisitAll(func(f *flag.Flag) {
+		envName := strings.ReplaceAll(strings.ToUpper(f.Name), "-", "_")
+		_, ok := os.LookupEnv(envName)
+		o.setFlags[f.Name] = ok || cliFlags.Has(f.Name)
+	})
+
+	if err := o.Validate(); err != nil {
+		return fmt.Errorf("validating options, %w", err)
+	}
+
+	// if ClusterID is not set, generate it from cluster endpoint
+	// if clusterEndpoint is empty, it might awaiting merge from MergeSettings(), do not run GetAPIServerName() or it could panic
+	// TODO: chore: remove clusterEndpoint validation logic here when karpenter-global-settings (and merge logic) are completely removed
+	if o.ClusterID == "" && o.ClusterEndpoint != "" {
+		o.ClusterID = getAKSClusterID(o.GetAPIServerName())
+	}
+
+	return nil
+}
+
+func (o *Options) ToContext(ctx context.Context) context.Context {
+	return ToContext(ctx, o)
+}
+
+func (o *Options) MergeSettings(ctx context.Context) {
+	s := settings.FromContext(ctx)
+	mergeField(&o.ClusterName, s.ClusterName, o.setFlags["cluster-name"])
+	mergeField(&o.ClusterEndpoint, s.ClusterEndpoint, o.setFlags["cluster-endpoint"])
+	mergeField(&o.VMMemoryOverheadPercent, s.VMMemoryOverheadPercent, o.setFlags["vm-memory-overhead-percent"])
+	mergeField(&o.ClusterID, s.ClusterID, o.setFlags["cluster-id"])
+	mergeField(&o.KubeletClientTLSBootstrapToken, s.KubeletClientTLSBootstrapToken, o.setFlags["kubelet-client-tls-bootstrap-token"])
+	mergeField(&o.SSHPublicKey, s.SSHPublicKey, o.setFlags["ssh-public-key"])
+	mergeField(&o.NetworkPlugin, s.NetworkPlugin, o.setFlags["network-plugin"])
+	mergeField(&o.NetworkPolicy, s.NetworkPolicy, o.setFlags["network-policy"])
+	mergeField(&o.NodeIdentities, s.NodeIdentities, o.setFlags["node-identities"])
+
+	if err := o.validateRequiredFields(); err != nil {
+		panic(fmt.Errorf("checking required fields, %w", err))
+	}
+
+	// if ClusterID is not set, generate it from cluster endpoint
+	if o.ClusterID == "" {
+		o.ClusterID = getAKSClusterID(o.GetAPIServerName())
+	}
```

Comment:
would there ever by any difference between o.ClusterID set in env vs generating like this?

Reply:
In this way, a part of generated ClusterID is randomized. Users can choose to customize their own string if they don't want the randomization.

Reply:
I don't think users should be able to provider their own string, it needs to be accurate to reference the fqdn. 

https://github.com/Azure/karpenter-provider-azure/blob/main/Makefile-az.mk#L47 it doesn't look like we pass this value in from flags or env variables. We probably should remove the if o.clusterId == "" because we need this value to be accurate always and shouldn't accept user input.

 Its used for a critical value https://github.com/Azure/karpenter-provider-azure/blob/main/pkg/providers/imagefamily/bootstrap/aksbootstrap.go#L452 so probably best to calculate it ourselves.

Reply:
@tallaxes  thoughts?

Reply:
@tallaxes I think the consideration of changing this should be in a separate PR or issue, regardless of whether it should be removed or not. This proposed change in ClusterID is too far away from the purpose of this PR--to migrate from Settings to Options.
Unless it the consideration small enough like the unit test changes earlier...

Reply:
Agree with both :-) We should be using calculated ClusterID exclusively; the only reason it is exposed as a configuration setting is because until certain stage I did not have enough confidence it is always calculated correctly. Since in current scenarios it is never passed externally, the configuration setting should be safe to remove. Can be handled in a separate PR.

Reply:
@tallaxes The issue @Bryce-Soghigian brought up with the idea of separating PR is also: the separate PR will become a `TODO` that no one might pick up, and then never get fixed. 
My proposal is to create an issue to track this rather than letting it be a stray `TODO`, what do you think? I also think that that should only applied to not-so-small issues (we have to guess the size I think), otherwise it can be fixed along the way in the same PR.

Reply:
Totally agree with creating issues for pretty much any identified follow-up work.

Reply:
https://github.com/Azure/karpenter-provider-azure/issues/130 I will get back to this after this one.